File: codes/bigdata-learn/flink/flink-scala/src/main/scala/cc/laop/flink/stream/WaterMarkTest.scala
package cc.laop.flink.stream

import org.apache.flink.streaming.api.scala._
import org.apache.flink.streaming.api.windowing.assigners.TumblingProcessingTimeWindows
import org.apache.flink.streaming.api.windowing.time.Time

/**
 * @Auther: Administrator
 * @Date: create in 2021/1/22 17:27
 * @Description:
 */
object WaterMarkTest {

  def main(args: Array[String]): Unit = {

    val env = StreamExecutionEnvironment.getExecutionEnvironment

    env.getConfig.setAutoWatermarkInterval(10000)

    val ds = env.addSource(new KafkaConsumerFactory().build("watermarkTest01"))

    ds
      .map(it => {
        val sarr = it.split(",")
        Access(sarr(0).toInt, sarr(1).toLong)
      })
      .assignAscendingTimestamps(_.time)
      .windowAll(TumblingProcessingTimeWindows.of(Time.seconds(10)))
      .apply[Access]((tw, its, col) => {
        for (elem <- its) {
          col.collect(elem)
        }
      })
      .print()

    env.execute()
  }

}

case class Access(id: Int, time: Long)

File: codes/bigdata-learn/flink/flink-java/src/main/java/cc/laop/flink/java/stream/WatermarkTest.java
package cc.laop.flink.java.stream;

import org.apache.flink.api.common.eventtime.WatermarkStrategy;
import org.apache.flink.api.common.serialization.SimpleStringSchema;
import org.apache.flink.api.common.typeinfo.Types;
import org.apache.flink.api.java.tuple.Tuple3;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.windowing.assigners.TumblingEventTimeWindows;
import org.apache.flink.streaming.api.windowing.time.Time;
import org.apache.flink.streaming.connectors.kafka.FlinkKafkaConsumer;

import java.time.Duration;
import java.util.Properties;

/**
 * @Auther: Administrator
 * @Date: create in 2021/1/26 16:42
 * @Description:
 */
public class WatermarkTest {

    public static void main(String[] args) throws Exception {

        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        // env.getConfig().setAutoWatermarkInterval(10000);
        env.setParallelism(1);

        Properties prop = new Properties();
        prop.put("bootstrap.servers", "192.168.1.30:9092");
        prop.put("group.id", "test004");

        DataStreamSource<String> ds = env.addSource(new FlinkKafkaConsumer<>("watermarkTest01",
                new SimpleStringSchema(), prop));

        WatermarkStrategy<Tuple3<Integer, Long, Integer>> ws =
                WatermarkStrategy
                        .<Tuple3<Integer, Long, Integer>>forBoundedOutOfOrderness(Duration.ofSeconds(10))
                        .withTimestampAssigner((element, recordTimestamp) -> element.f1);

        ds
                .map(it -> {
                    Integer i = Integer.valueOf(it.substring(0, it.indexOf(",")));
                    Long time = Long.valueOf(it.substring(it.indexOf(",") + 1));
                    return new Tuple3<>(i, time, 1);
                })
                .returns(Types.TUPLE(Types.INT, Types.LONG, Types.INT))
                .assignTimestampsAndWatermarks(ws)
                .keyBy(it -> it.f0)
                .window(TumblingEventTimeWindows.of(Time.seconds(10)))
                // .window(SlidingEventTimeWindows.of(Time.seconds(10), Time.seconds(8)))
                .reduce((t1, t2) -> {
                    System.out.println("t1: " + t1);
                    System.out.println("t2: " + t2);
                    return new Tuple3<>(t2.f0, t2.f1, t2.f2 + t1.f2);
                })
                .print()
        ;

        env.execute();

    }


}
